### PR TITLE
build: enable ASAN

### PR DESCRIPTION
**Pull request by ingwinlu**
# Purpose

Enables a docker ASAN build to replace the current one

# Checklist

- [ ] commit messages are fine ("module: short statement" syntax and refer to issues)
- [ ] affected documentation is fixed
- [ ] release notes are updated (doc/news/_preparation_next_release.md)

### CODE REVIEW

**Comment by markus2330** [comment]
I think we need to disable lua for asan in this constellation then.

**Comment by ingwinlu** [comment]
@markus2330 just the test or the whole lua binding?

**Comment by markus2330** [comment]
If the plugin crashes it makes sense to remove it completely (in this situation with ASAN).

We already have code to exclude plugins for ASAN builds, e.g. for yamlcpp.

@sanssecours Why does the lua plugin disable tests for "APPLE AND ENABLE_ASAN"? Does the plugin generally work with ASAN and only the tests fail?

**Comment by sanssecours** [comment]
> Why does the lua plugin disable tests for "APPLE AND ENABLE_ASAN"?

As far as I remember because the tests do not work when ASAN is enabled (AddressSanitizer is loaded too late).

> Does the plugin generally work with ASAN and only the tests fail?

Seems the **bindings** work fine. At least `require kdb` works inside the Lua interpreter after I export `LUA_CPATH` and set `DYLD_INSERT_LIBRARIES` according to the instruction printed after I used `require "kdb"` without setting `DYLD_INSERT_LIBRARIES`:

```
==80684==ERROR: Interceptors are not working. This may be because AddressSanitizer is loaded too late (e.g. via dlopen). Please launch the executable with:
DYLD_INSERT_LIBRARIES=/usr/local/opt/llvm/lib/clang/6.0.0/lib/darwin/libclang_rt.asan_osx_dynamic.dylib
```

.

**Comment by markus2330** [comment]
I disabled https://build.libelektra.org/jenkins/job/elektra-clang-asan/

**Comment by ingwinlu** [comment]
I am still not sure how you want me to proceed here.

**Comment by markus2330** [comment]
We should run ASAN with clang and gcc in the Jenkinsfile. If there are errors, we disable the respective plugin (in the constellation where it causes problem), document the reduced test coverage in doc/todo/TESTING (or if there is a good chance that someone fixes it by GitHub Issue).

The lua problems are unlikely to be fixed soon, as currently nobody seems to need this plugin.

**Comment by ingwinlu** [comment]
Did you look at the list of failing tests?
```
The following tests FAILED:
	 33 - testhaskell_basic (Failed)
	 34 - testhaskell_basic_optimized (Failed)
	 35 - testhaskell_realworld (Failed)
	 36 - testhaskell_realworld_optimized (Failed)
	 37 - testpy2_kdb.py (Failed)
	 38 - testpy2_key.py (Failed)
	 39 - testpy2_keyset.py (Failed)
	 40 - test_kdb.py (Failed)
	 41 - test_key.py (Failed)
	 42 - test_keyset.py (Failed)
	 43 - test_kdb.lua (Failed)
	 44 - test_key.lua (Failed)
	 45 - test_keyset.lua (Failed)
	 46 - testruby_kdb (Failed)
	 47 - testruby_key (Failed)
	 48 - testruby_keyset (Failed)
	 49 - testruby_tools_backendbuilder (Failed)
	 50 - testruby_tools_backendparser (Failed)
	 51 - testruby_tools_modules (Failed)
	 52 - testruby_tools_plugindatabase (Failed)
	 53 - testruby_tools_pluginspec (Failed)
	 54 - testruby_tools_specreader (Failed)
	 55 - testglib_kdb (Failed)
	 56 - testglib_key (Failed)
	 57 - testglib_keyset (Failed)
	 62 - testio_glib (Failed)
	 65 - testshell_markdown_base64 (Failed)
	 80 - testmod_crypto_botan (Failed)
	103 - testshell_markdown_ini (Failed)
	113 - testmod_lua (Failed)
	117 - testshell_markdown_mini (Failed)
	129 - testshell_markdown_tcl (Failed)
	149 - testscr_check_kdb_internal_check (Failed)
	177 - testshell_markdown_tutorial_validation (Failed)
```
If we disable all those constellations we might as well not run it as all.

**Comment by markus2330** [comment]
No sorry, I did not have time to look at it.

But nearly everything seems to be related to swig (haskell, python, ruby) and glib. We can also disable all swig/glib bindings for this ASAN build. The other parts are actively maintained:

- testmod_crypto_botan and testshell_markdown_base64 by @petermax2 
- testshell_markdown_mini testshell_markdown_ini by @sanssecours 
- testshell_markdown_tutorial_validation is known to fail, also without ASAN, see #1926

If we continue to not run it (or run it with errors which means the build is ignored) the code will degrade even more. If we check for what works now, we at least do not get regressions.

As long as there is no upstream support of ASAN builds for our dependencies (and it seems to be lacking for quite some libs and languages) it is unrealistic that everything works with ASAN. This is not something that can be fixed within our source.

But we as library should enable other applications to be ASAN and valgrind clean. So we need to be pioneers in this area.

**Comment by sanssecours** [comment]
Concerning `testshell_markdown_mini` the test works fine on Ubuntu (using [GCC](https://travis-ci.org/ElektraInitiative/libelektra/jobs/378829078#L1969) and [Clang](https://travis-ci.org/ElektraInitiative/libelektra/jobs/378829079#L1707)) and macOS (using [Clang](https://travis-ci.org/ElektraInitiative/libelektra/jobs/378829077#L6995)). 

The test fails on my machine, if I enable the leak sanitizer. I can fix the problem by not requiring the `code` plugin though. So I guess the real culprit here is the `code` plugin, since requiring other plugins (I tried `mathcheck`) seems to work fine.

**Comment by markus2330** [comment]
Thank you for looking into it!

> I am pretty certain that you misspelled @tom-wa here. 

Okay, then this test needs to be excluded, too :wink: Luckily INI is not the default.

> I can fix the problem by not requiring the code plugin though.

Sounds good. Alternatively you could try one of the other code plugins?

I am also okay with excluding all these tests for this ASAN build, none of them tests essential or core functionality.

**Comment by ingwinlu** [comment]
```
-- Exclude Binding glib because glib is not compatible with ENABLE_ASAN
CMake Error at src/bindings/gsettings/CMakeLists.txt:35 (add_subdirectory):
  The binary directory
    /home/jenkins/workspace/libelektra_PR-1963-UY7VIGJMXZ2D2SSXYCECW7I6XLHKVOJM6XQS67PEYRQNRNHU5NRA@3/build/src/bindings/glib
  is already used to build a source directory.  It cannot be used to build source directory
    /home/jenkins/workspace/libelektra_PR-1963-UY7VIGJMXZ2D2SSXYCECW7I6XLHKVOJM6XQS67PEYRQNRNHU5NRA@3/src/bindings/glib
  Specify a unique binary directory name.
```
Is there no handling of binding dependencies at all?

**Comment by markus2330** [comment]
The dependency handling needs to be implemented manually in CMake. Seems like it was forgotten in the (unfortunately unmaintained) gsettings.

Please implement it or at least open separate bug reports.

> Is there no handling of binding dependencies at all?

A very short time ago it was not even easily possible to query which binding is included. It was a lot of work done recently by @waht.

**Comment by ingwinlu** [comment]
jenkins build libelektra please

File: cmake/Modules/LibAddHaskellPlugin.cmake [hunk 1]
```diff
@@ -233,28 +254,6 @@ macro (add_haskell_plugin target)
 			remove_plugin (${target} "${PLUGINPROCESS_NOTFOUND_INFO}, but required for haskell plugins")
 		endif (PLUGINPROCESS_FOUND)
 	endif (DEPENDENCY_PHASE)
-
-	# compile our c wrapper which takes care of invoking the haskell runtime
-	# the actual haskell plugin gets linked in dynamically as a library
-	add_plugin (${target}
```
Comment (markus2330) [inline cmake/Modules/LibAddHaskellPlugin.cmake]: add_plugin needs to be present at all stages, to disable plugins remove_plugin needs to be called.

Comment (ingwinlu) [reply cmake/Modules/LibAddHaskellPlugin.cmake]: that was not the problem

**Comment by ingwinlu** [comment]
@markus2330 i can not add codeblocks for in code comments so I will answer here in full

```
-- Configuring done
CMake Error at cmake/Modules/LibAddPlugin.cmake:442(add_dependencies):
  The dependency target "c2hs_haskell" of target "elektra-haskell" does not
  exist.
Call Stack (most recent call first):
  cmake/Modules/LibAddHaskellPlugin.cmake:239 (add_plugin)
  src/plugins/haskell/CMakeLists.txt:3 (add_haskell_plugin)
CMake Error at cmake/Modules/LibAddPlugin.cmake:402 (add_dependencies):
  The dependency target "c2hs_haskell" of target "elektra-haskell-objects"
  does not exist.
Call Stack (most recent call first):
  cmake/Modules/LibAddHaskellPlugin.cmake:239 (add_plugin)
  src/plugins/haskell/CMakeLists.txt:3 (add_haskell_plugin)
```

As you can see the block needed to be moved up to where all the if blocks are. And that is all that I changed.

**Comment by ingwinlu** [comment]
@markus2330 any pointers on how to handle the remaining failures: https://build.libelektra.org/jenkins/blue/organizations/jenkins/libelektra/detail/PR-1963/34/pipeline/274 they don't quite seem to be as optional as the earlier tests i disabled

**Comment by sanssecours** [comment]
I think I also know how to “fix” `testshell_markdown_tutorial_validation`. I have to go to work now though. I will probably push an update for `testshell_markdown_tutorial_validation` later.

File: src/plugins/ini/README.md [hunk 2]
```diff
@@ -3,7 +3,7 @@
 - infos/licence = BSD
 - infos/needs =
 - infos/provides = storage/ini
-- infos/recommends = binary
+- infos/recommends = base64
```
Comment (markus2330) [inline src/plugins/ini/README.md]: In the clauses recommends/needs we should be as abstract as possible (so that mount also works if base64 is not present, but some other binary-transformation plugin is present). It is better to rewrite the tutorials.

**Comment by sanssecours** [comment]
As far as I can tell “all” tests should now pass on the ASAN enabled build. For anyone interested, the following tests still fail on my machine (macOS, llvm 6):

- `testmod_type`
- `testtool_backendbuilder`
- `testtool_plugindatabase`
- `testmod_augeas`
- `testmod_dbus`
- `testshell_markdown_xerces`
- `testmod_xerces`
- `testscr_check_kdb_internal_check`
- `testkdb_allplugins`

.

Log: [macOS - ASAN.log.txt](https://github.com/ElektraInitiative/libelektra/files/2029739/macOS.-.ASAN.log.txt)

**Comment by ingwinlu** [comment]
Thanks for your help.
We can add another test with clang + asan enabled in this pr if you want to tackle those as well (or do it in a new one).

**Comment by sanssecours** [comment]
> We can add another test with clang + asan enabled in this pr if you want to tackle those as well (or do it in a new one).

Nope, thanks. I think I am good for now 😊. I should work on [something different](https://github.com/sanssecours/yan-lr) after all…

File: tests/shell/check_kdb_internal_check.sh [hunk 3]
```diff
@@ -33,6 +33,23 @@ do
 		;;
 	esac
 
+	# The following checks fail on an ASAN enabled build
+	# See also: https://github.com/ElektraInitiative/libelektra/pull/1963
```
Comment (markus2330) [inline tests/shell/check_kdb_internal_check.sh]: please document in doc/todo/TESTING

File: src/plugins/tcl/README.md [hunk 4]
```diff
@@ -2,7 +2,7 @@
 - infos/author = Markus Raab <elektra@libelektra.org>
 - infos/licence = BSD
 - infos/provides = storage/tcl
-- infos/needs = code null binary
+- infos/needs = ccode null base64
```
Comment (markus2330) [inline src/plugins/tcl/README.md]: as already said, it would be better if we can use generic providers here (like binary) and fix the tutorials to use concrete plugins (that pass ASAN checks)

File: src/plugins/ini/README.md [hunk 5]
```diff
@@ -146,7 +146,7 @@ By default the INI plugin does not support binary data. You can use the [Base64
 
 ```sh
 # Mount INI and recommended plugin Base64
-sudo kdb mount --with-recommends config.ini user/examples/ini ini
+sudo kdb mount config.ini user/examples/ini ini base64
```
Comment (markus2330) [inline src/plugins/ini/README.md]: Doesn't --with-recommends  together with base64 also do the right thing? And we should document that base64 is added to pass ASAN tests.

File: src/bindings/swig/CMakeLists.txt [hunk 6]
```diff
@@ -6,6 +6,11 @@ if (NOT SWIG_FOUND)
 	return ()
 endif ()
 
+if (ENABLE_ASAN)
+	exclude_binding (swig "SWIG is not compatible with ENABLE_ASAN")
+	return ()
```
Comment (markus2330) [inline src/bindings/swig/CMakeLists.txt]: Would this return be problematic once we change the bindings to behave like the plugins?

File: src/bindings/glib/CMakeLists.txt [hunk 7]
```diff
@@ -3,40 +3,48 @@ include (LibAddMacros)
 find_package (PkgConfig)
 pkg_check_modules (GOBJECT gobject-2.0>=2.36)
 
+# Test if binding can be build
 if (NOT GOBJECT_FOUND)
 	exclude_binding (glib "No gobject found using pkg-config, also remove gi bindings because they depend on it")
-else ()
-	add_binding (glib)
+	return ()
```
Comment (markus2330) [inline src/bindings/glib/CMakeLists.txt]: See below. Furthermore the diff is hard to read because of this reformatting.

File: cmake/Modules/LibAddPlugin.cmake [hunk 8]
```diff
@@ -52,6 +52,10 @@ function (add_plugintest testname)
 						"Trying to add plugintest ${testname} but no such plugin was added (try to use LINK_PLUGIN)"
 					)
 			endif ()
+
+			# plugin for test was not added in previous phase or removed
```
Comment (markus2330) [inline cmake/Modules/LibAddPlugin.cmake]: Is this also relevant for the docu above?

File: cmake/Modules/LibAddHaskellPlugin.cmake [hunk 9]
```diff
@@ -215,6 +215,78 @@ macro (add_haskell_plugin target)
 									     "${GHC_GMP_PATH}"
 									     "${CMAKE_INSTALL_RPATH}"
 									     "${CMAKE_INSTALL_PREFIX}/lib${LIB_SUFFIX}/elektra/haskell")
+
```
Comment (markus2330) [inline cmake/Modules/LibAddHaskellPlugin.cmake]: @e1528532 Can you take a look?

This certainly needs a restructuring.

File: src/plugins/tcl/README.md (hunk 4 above)
Comment (ingwinlu) [reply src/plugins/tcl/README.md]: Would it make sense to solve this in a different PR?

File: src/bindings/swig/CMakeLists.txt (hunk 6 above)
Comment (ingwinlu) [reply src/bindings/swig/CMakeLists.txt]: I don't expect so

File: tests/shell/check_kdb_internal_check.sh (hunk 3 above)
Comment (ingwinlu) [reply tests/shell/check_kdb_internal_check.sh]: done in 90a66087419867be2750445fde8636a6ebf57f11

File: src/plugins/ini/README.md (hunk 5 above)
Comment (sanssecours) [reply src/plugins/ini/README.md]: > Doesn't --with-recommends together with base64 also do the right thing?

While this works, using `--with-recommends` is redundant and might also be confusing here.

File: src/plugins/tcl/README.md (hunk 4 above)
Comment (sanssecours) [reply src/plugins/tcl/README.md]: > we can use generic providers here (like binary) and fix the tutorials to use concrete plugins

I do not think that will work here, since these plugins are listed under `infos/needs` and not under `infos/recommends`.

Comment (markus2330) [reply src/plugins/tcl/README.md]: If you explicitly state needed plugins, kdb mount will know that the dependency is already fulfilled.

File: src/plugins/ini/README.md (hunk 5 above)
Comment (markus2330) [reply src/plugins/ini/README.md]: Yes, thus we should explain that "base64" is only there to make ASAN happy.

**Comment by sanssecours** [comment]
jenkins build libelektra please

**Comment by ingwinlu** [comment]
anything else @markus2330?

**Comment by markus2330** [comment]
jenkins build all please

**Comment by markus2330** [comment]
Thank you, looks good!

I am not sure if the kdb mount itself leaks memory or later commands but I think the added lines should be clear anyway.

> anything else @markus2330?

You can simply add "ready to merge" if you finished the reviews and all build jobs finished successfully. (Or additionally ping me if I do not respond).

Btw. if you have changes beyond Jenkinsfile, please do not forget to call "jenkins build all please".

**Comment by sanssecours** [comment]
jenkins build libelektra please

**Comment by ingwinlu** [comment]
jenkins build gcc-configure-debian-optimizations please
